src/v/coproc/materialized_topics_frontend.cc

```diff
@@ -0,0 +1,106 @@
+          result.tp_ns);
+        if (result.ec == cluster::errc::success) {
+            vlog(
+              coproclog.info, "Materialized topic created: {}", result.tp_ns);
+            found->second->set_value();
+        } else if (result.ec == cluster::errc::topic_already_exists) {
+            vlog(
+              coproclog.info,
+              "Materialzed log has come into existance via "
+              "another node: {}",
+              result.tp_ns);
+            found->second->set_value();
+        } else {
+            found->second->set_exception(
+              materialized_topic_replication_exception(fmt::format(
+                "Failed to replicate materialized topic: {}, error: {}",
+                result.tp_ns,
+                result.ec)));
+        }
+        _topics.erase(found);
+    }
+}
+
+void materialized_topics_frontend::topic_creation_exception(
+  const std::vector<cluster::non_replicable_topic>& topics,
+  const std::exception& ex) {
+    for (const auto& topic : topics) {
+        auto found = _topics.find(topic.name);
+        if (found != _topics.end()) {
+            found->second->set_exception(
+              materialized_topic_replication_exception(fmt::format(
+                "create_non_replicable_topics threw: {}, when attempting to "
+                "create non_replicable_topic: {}",
+                ex.what(),
+                topic)));
+        } else {
+            vlog(
+              coproclog.error,
+              "Exception thrown for materialized topic creation that doesn't "
+              "exist in requests cache: {}",
+              topic);
+        }
+        _topics.erase(found);
+    }
+}
+
+ss::future<> materialized_topics_frontend::create_materialized_topics(
+  std::vector<cluster::non_replicable_topic> topics,
+  model::timeout_clock::time_point timeout) {
+    std::vector<ss::future<>> all;
+    std::vector<cluster::non_replicable_topic> todos;
+    for (auto& topic : topics) {
+        auto [itr, success] = _topics.try_emplace(
+          topic.name, ss::make_lw_shared<ss::shared_promise<>>());
+        if (!success) {
+            all.emplace_back(itr->second->get_shared_future());
+        } else {
+            todos.emplace_back(std::move(topic));
+        }
+    }
+    if (!todos.empty()) {
+        auto f = _topics_frontend.local()
+                   .create_non_replicable_topics(todos, timeout)
+                   .then(
+                     [this](const std::vector<cluster::topic_result>& result) {
+                         topic_creation_resolved(result);
+                     })
+                   .handle_exception_type([this, todos = std::move(todos)](
+                                            const std::exception& ex) {
+                       topic_creation_exception(todos, ex);
+                   });
+        all.emplace_back(std::move(f));
+    }
+    return ss::when_all_succeed(all.begin(), all.end());
+}
```

Comment:
this looks to me like it belongs in `src/v/cluster/*` along with the other frontends, right? a common pattern would have the frontend sibling to the other frontends, and sub-system specific backends in coproc directory, as an example. @mmaslankaprv do you see value in consolidating or having frontends in separate subsystems is ok?

Reply:
To me, the only value this class adds is this debounce mechanism, and the reason it needs to be created is specific to its usage within coproc. So that's my reasoning for having this class in `v/coproc`, @mmaslankaprv opinions?

Reply:
many of the frontends don't do anything except error handling, debounce, serialization, routing, etc... and then they rely on sub-system specific things. data policy, security/users, acl, are examples.

i really don't have a strong preference, but we should value maintaining existing patterns so that people can form reliable intuitions about where things are located, even if there is no functional value.

Reply:
I really like the idea of moving this fronted to `cluster/*` module exactly as @dotnwat, suggested. This way we will have all the metadata related services in the same module.